### DISCUSSION TITLE
Can I define a custom source or folder from which my site hosted on GitHub pages can load from?

### DISCUSSION THREAD

**Discussion by osioke** [community/community#23073] [Other Feature Feedback, Questions, & Ideas]
<p>Instead of the normal options as the source for my site deployed with pages, that is the <code>master</code> branch or the <code>docs</code> folder in the master branch, can I define another source, like a folder called <code>site</code> or <code>pages</code> or some custom name.</p>
<p>This will be really helpful, as I really like using specific names for my repo folders.</p>

**Comment by winstliu** [comment, marked as answer]
<p>From <a href="https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/" rel="nofollow noopener">https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/</a>, the only three options that GitHub Pages recognizes are:</p>
<ul>
<li>master branch</li>
<li>docs/ folder on the master branch</li>
<li>gh-pages branch</li>
</ul>
<p>So it doesn’t look like you’re able to define a custom folder or branch yet.</p>

**Reply by Paul-Browne** [reply]
you can now specify the branch. Doesn't have to be named gh-pages

**Reply by a0ax** [reply]
@Paul-Browne How? Is there anywhere I can read more?

**Reply by rayluo** [reply]
FWIW, this is the solution [Github Page Overwriter](https://github.com/marketplace/actions/github-pages-overwriter) I came up with and have been using.

**Reply by Andre601** [reply]
I legit don't get the point of that action tho... What is the problem with hosting your website's files on the repo (if I understand it right)?
You can keep it in a separate branch and have it build when pushing to it, before deploying.

I can't see where it "polutes" the git history here. If anything does that even help keeping things transparent.

**Reply by airtower-luna** [reply]
You can also build and deploy a page without pushing the build result now, see: https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-with-a-custom-github-actions-workflow

**Reply by rayluo** [reply]
> I legit don't get the point of that action tho... What is the problem with hosting your website's files on the repo (if I understand it right)? You can keep it in a separate branch and have it build when pushing to it, before deploying.
> 
> I can't see where it "polutes" the git history here. If anything does that even help keeping things transparent.

@Andre601 , we already had similar conversation before. Please refer to [this message](https://github.com/orgs/community/discussions/23073#discussioncomment-3238961) and the 2 more after it. Your mindset seems remain unchanged "there is already an X, why invent a new Y". You just replaced your X1 in 2021 with a different X2 today. Well, probably the inventor of Y [saw something lacking in X](https://github.com/orgs/community/discussions/23073#discussioncomment-3238963). And it is also possible that you need no more feature than X, in that case you do not feel a need to switch, which is totally fine.

The choice criteria is simple. If your repo's file structure already matches what you want to show as a website (for example, you already have an `index.html` in your root folder), you do not need any extra github actions. Just deploy it as-is to a github page, and you are all set. This github discussion started in 2017 for a reason, that the `index.html` file either is inside a sub folder, or does not even exist until an artifact is built. Now, in all fairness, that feature might have been satisfied by recent improvements in 2023 (although I still can't find it in luna's doc link above). As a user you can always shop around and pick the solution that suit you. I'm merely providing an option here.

P.S.: If you want to continue our discussion, please reply to that [earlier message](https://github.com/orgs/community/discussions/23073#discussioncomment-3238963) where we left off. I do not want to hijack this chosen answer link and "pollute" it.

**Comment by osioke** [comment]
<p>Yea <a class="mention" href="/50wliu">@50wliu</a>, true. Thank you for helping out, I am grateful.</p>

**Comment by tajmone** [comment]
<p>I agree with <a class="mention" href="/osioke">@osioke</a>, being able to choose a custom folder name would be great — I was about to create a post on this topic, but since I’ve found this thread I just joined in, so I’ll share here some thoughts on this.</p>
<p>I find that having only the “docs/” folder as an option for serving GH-Pages contents directly within the main repository is very limiting. The idea of hosting the GH-Pages contents inside the repo is a <em>really cool</em> feature, which I use a lot for online documentation, since it allows smart reuse of the repository contents to generate the documentation without having to work across orphan branches.</p>
<p>But in most cases, I find the “docs” folder name problematic, for I’d rather use that folder for the software/project documentation, whereas for the GH-Pages contents a folder named “www” would have been more intuitive and practical  — or at least the freedom to rename the folder as I please.</p>

**Reply by itsalaidbacklife** [reply]
I would love the flexibility to specify any directory of the project, and I agree that an alternative top-level predefined name which was more clear about intent would be preferable to `docs`. I suggest `gh-pages` as the folder name to match the predefined branch pattern, and to be even more explicit than www

**Comment by webdott** [comment]
<p><a class="mention" href="/50wliu">@50wliu</a> I do not have gh-pages branch. Only the ‘master’ and ‘docs/folder’ branch. Is there anything I can do about this?</p>

**Comment by Andre601** [comment]
<aside class="quote no-group quote-modified" data-username="50wliu" data-post="2" data-topic="10237" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 50wliu:</div>
<blockquote>
<p>From <a href="https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/" rel="noopener nofollow ugc">https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/ </a>, the only three options that GitHub Pages recognizes are:</p>
<ul>
<li>master branch</li>
<li>docs/ folder on the master branch</li>
<li>gh-pages branch</li>
</ul>
<p>So it doesn’t look like you’re able to define a custom folder or branch yet.</p>
</blockquote>
</aside>
<p>There is also a special case for user pages (Those directly available under <code>:username.github.io</code>) which is, that they only accept the <code>master</code> branch as an option. Every other option is not selectable.</p>

**Comment by Andre601** [comment]
<p>(Bit late on a response but better late then never I guess?)</p>
<p>With GitHub Actions could you setup an action, that would listen for pushes towards a specific folder (your <code>www</code> folder for example), take the content from it, build it, if nessesary and then pushes it towards the <code>gh-pages</code> branch.<br>
That way could you have your custom folder while still using what GitHub offers.</p>

**Comment by tajmone** [comment]
<p><a class="mention" href="/Andre601">@Andre601</a>, I like your suggestion of using GitHub Actions to publish on the  <code>gh-pages</code>  branch — it also allows to spare having to host the HTML files in the source repo (in case of docs being built from a lightweight markup syntax).</p>
<p>But I still think that in some case it might be desirable to have the HTML files inside the project, e.g. when the documentation is intended for both online and offline reading.</p>
<p>Being able to customize the GHPages folder name would be quite practical, and would allow to accommodate the user’s naming convention — e.g. <code>www</code> for the built HTML docs, and <code>www-src</code> for the markdown/AsciiDoc source, etc.</p>
<p>Mainly, I think that naming the default GHPages folder <code>docs</code> was not an ideal choice, for many projects would rather use that folder name to host documentation, whereas the website might be just an introductory page presenting the project.</p>
<p>If the folder was named <code>gh-pages</code> in the first place,  this wouldn’t be an issue, but since many GitHub repositories are also hosted on other platforms (GitLab, Bitbucket, etc.) it would probably be too restrictive — hence the proposal of using <code>www</code> instead, which is a common folder name used on LAMP/WAMP stacks (at least as an alias).</p>

**Comment by Thaina** [comment]
<p>I want to have this as feature request. We should allowed to specified any folder of the repo as pages</p>
<p>From my used case, I have create my web repo with blazor. And the default folder it use for page is <code>wwwroot</code>. Which, if github pages allowed custom folder specified, I could specify github pages with <code>wwwroot</code> folder and just create my repo as that</p>

**Comment by mluevanos** [comment]
<p>Based on <a href="https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/" rel="noopener nofollow ugc">https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/</a> I am not sure if gh-pages branch is still an option.</p>
<p>I found: <a href="https://gist.github.com/cobyism/4730490" class="inline-onebox" rel="noopener nofollow ugc">Deploy to `gh-pages` from a `dist` folder on the master branch. Useful for use with [yeoman](http://yeoman.io). · GitHub</a> which is " Deploying a subfolder to GitHub Pages"</p>
<ul>
<li>I did not use this, I ended up moving my files to the root, but the discussion is still active</li>
<li>Next time I publish a site I will probably just try to create a different branch with my index.html in there and select the branch from drop down as the publishing source. GitHub documentation/processes changes so much it’s difficult to know what will work!</li>
</ul>

**Comment by Andre601** [comment]
<p>The <code>gh-pages</code> branch is still a thing.<br>
In fact will GitHub automatically enable GitHub Pages, if a <code>gh-pages</code> branch was created, so it’s still supported.<br>
I myself use this branch for various projects where I have the docs or similar on <a href="http://github.io" rel="noopener nofollow ugc">github.io</a></p>

**Comment by rayluo** [comment]
<p><a class="mention" href="/tajmone">@tajmone</a> , <a class="mention" href="/Andre601">@Andre601</a> , <a class="mention" href="/mluevanos">@mluevanos</a> , et al, I encountered the same need, and - it took me much longer than expected but - I finally came up with a reusable Github Actions named <a href="https://github.com/marketplace/actions/github-pages-overwriter" rel="noopener nofollow ugc">Github Pages Overwriter</a> to allow choosing arbitrary folder in your repo to publish it into Github Pages.</p>
<p>Give it a shot, spread the words if you like it, subscribe <a href="https://github.com/rayluo/github-pages-overwriter" rel="noopener nofollow ugc">my project</a> if you want to receive future updates, <a href="https://github.com/rayluo/github-pages-overwriter/issues" rel="noopener nofollow ugc">report issues here</a> because I may not come back and check this post.</p>

**Comment by Andre601** [comment]
<p>There is already <a href="https://github.com/peaceiris/actions-gh-pages" rel="noopener nofollow ugc">peaceiris/actions-gh-pages</a> which does this job.<br>
There, you can define your own folders from which the action would take the files to publish to the gh-pages branch.</p>

**Comment by rayluo** [comment]
<p>Hi <a class="mention" href="/Andre601">@Andre601</a> , good point. And, yes, <a href="https://github.com/peaceiris/actions-gh-pages" rel="noopener nofollow ugc">peaceires/actions-gh-pages</a> was (and still is) among the top search results when I was evaluating existing solutions. Here come the major reasons that I eventually started the <a href="https://github.com/marketplace/actions/github-pages-overwriter" rel="noopener nofollow ugc">Github Pages Overwriter</a> project.</p>
<ol>
<li>
<p>The What It Does. I prefer a git tenet of never store build artifacts inside a repo (at least not store as part of the permanent history). Some existing github actions may possibly satisfy that, but none of them explicitly promises such behavior in their contract (i.e. their doc). I tested peaceires/actions-gh-pages just now. It always adds a new commit into the <code>gh-pages</code> branch. Not the way I prefer.</p>
<p>In my new project, not only I name it “Github Page OVERWRITER” which hints that the generated content would be ephemeral and overwritten in next deployment, I also specifically <a href="https://github.com/marketplace/actions/github-pages-overwriter#how-does-it-work" rel="noopener nofollow ugc">documented that in “how does it work” section</a>.</p>
<p>In order to examine whether an existing Github Pages deployment action satisfies the aforementioned tenet, I’d need to look into their implementation. But, many of them turn out to be implemented by one programming language or another, running in some sort of container/runner. I certainly do not know all existing languages, so I cannot easily review any given 3rd-party github action. This brings me to my next point.</p>
</li>
<li>
<p>The How It Is Implemented. To satisfy tenet in <span class="hashtag">#1</span>, one only needs to run several git commands in the right way, without much external dependency. Feel free to look into the “Github Page Overwritter” source code to see how it only uses git command line tool to achieve that. Personally, it took me less time to write a new action to do what I want, then to sieve through existing actions to see whether one of them happens to suit my need.</p>
</li>
<li>
<p>The Security.  Any 3rd-party github action could potentially mess up with your central repo. <a href="https://github.community/t/are-github-actions-safe-to-use/17895">Security concern on github action is a real thing</a>. The only way to be sure, is to review their implementation. This is another reason why “<span class="hashtag">#2</span> how it is implemented” is important. I feel more comfortable to trust an action that I can fully understand what it works. So should you.</p>
</li>
<li>
<p>The Simplicity. Based on <span class="hashtag">#1</span> and <span class="hashtag">#2</span>, there is really no much to configure. A simpler configuration means easier to get started.</p>
<p>FWIW, Github Page Ovewriter does not require any github token to be configured. My attempts with the “Actions-gh-pages”, on the other hand, failed with "Error: Action failed with “not found deploy key or tokens” despite that the “github_token” is documented as optional in its doc.</p>
</li>
</ol>
<p>Perhaps I should also include this essay into my project’s README file. :slight_smile:</p>

**Comment by tajmone** [comment]
<p>In the meantime, since this discussion took place, GitHub added a new option that allows using a custom-named branch (not just <code>gh-branch</code>).</p>

**Comment by dangowans** [comment]
<p>I would love the option of a <code>www</code> folder!</p>

**Comment by seth127** [comment]
<p>This has been going on for awhile with not much movement, but I’ll drop in to echo the sentiment: I would love to be able to point this to any arbitrary folder on any branch!</p>
<p>I often use github pages for previewing a site before pushing it to its public destination, so being able to point to anywhere within the repo would be very useful.</p>
<p>Also, I may be wrong, but it looks to me like you can currently point to <em>any branch</em> (not just <code>master</code> or <code>gh-pages</code>) but you still only have the option of pointing to the root or a <code>docs/</code> dir.</p>

**Comment by jeffnyman** [comment]
I'll throw my hat into the ring saying that at least some other options, if not outright configuration of any directory, would be nice here.

The use of `docs` feels like an odd choice since many projects would have that folder reserved for documentation. Why `www` or `wwwroot` or `site` or `public` or -- well, really _any_ one of the many other well-established root names for a web site -- wasn't chosen isn't clear to me.

There might be very good reasons why `docs` and only `docs` is allowed but, if so, it would at least be good to hear the rationale for that.

**Reply by Andre601** [reply]
GitHub Pages deployment are now using actions, so it can be configured more freely.

**Reply by jeffnyman** [reply]
Got it. Yeah, I was seeing that focus on Actions as I searched. I'm not sure if I'm seeing a simple action that lets me just use `public`, for example, without having to do other steps, such as `gh-pages` or branches, since I'm not doing any of that. I just have a `main` branch that has a `public` directory and that latter directory will hold what should be served.

**Reply by airtower-luna** [reply]
There's a starter workflow that uploads a static page, adjusting the [path used for `actions/upload-pages-artifact`](https://github.com/actions/starter-workflows/blob/e03baab3aad7266e9a3c8658e2bf4c86c9a1639d/pages/static.yml#L39) should be enough. Plus the branch filter for the push event of course, and you might want to [add an environment protection rule](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) to enforce that filter.

**Reply by jeffnyman** [reply]
Much thanks! This did the exact trick. I clearly just needed to be more involved with GitHub Actions it seems.

**Comment by auwsom** [comment]
Forgive my cynicism, but this seems silly. There is a dropdown for directories that could easily populate dirs other than /root or /docs. 
I agree limiting to docs as a folder is bizarre for the current prevalence of static sites. 
I would use a symlinked locally to a docs folder, but it doesnt work as git sees symlinks as files.
Actions seems like overkill for a common configuration of having a subdirectory for static site generation.
I tried this Gist, which creates a GithubPages branch: https://gist.github.com/cobyism/4730490 "Deploying a subfolder to GitHub Pages", but it looks like I would have to do this for every update.

KISS method: symlink the directory to another outside of the local repo and create a new repo for that directory. `git init && git remote add origin git@github.com:<username>/<repo>.git`

**Reply by Andre601** [reply]
Why not use the new option of GitHub Actions to push to pages?
GitHub implemented this option pretty much for this: To allow a lot more freedom in your setup.

**Reply by wsshin** [reply]
> I would use a symlinked locally to a docs folder, but it doesnt work as git sees symlinks as files.

symlink works for me.  Suppose your static site generator generates the output files in a specific folder, say `website/`.  I think you probably did `ln -s website docs`.  This doesn't work, because `git` sees the symbolic link `docs` as a file as you pointed out.

You should reverse the direction of the symlink: `ln - s docs website`.  Then, when the static site generator generates files in `website/`, actually `docs/` is populated.  Now you can push the populated `docs/` to GitHub to deploy the website.

This may not be an ideal solution, but it is a method that works for me.

**Comment by auwsom** [comment]
Ah, yes, the reverse would work. Thanks!

Also, symlink the output dir a level up and create a new repo for it. This
is what I've chosen as the separation of the generator with the site repos
makes sense to me. I can gitignore the site output, and dev on the
generator, while having CICD on the site repo when I edit/blog on it.

**Comment by katbyte** [comment]
+1 for a custom path, limiting it to `/docs` and the root makes little sense

**Comment by Navil** [comment]
Is there any plan of implementing this in the near future? I feel like writing actions for such a basic configuration is an overkill.

**Reply by airtower-luna** [reply]
This has been implemented for a few months now, since you can use a custom deployment workflow: https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-with-a-custom-github-actions-workflow

**Reply by Navil** [reply]
Yes, but only by writing a GitHub actions and taking care of the workflow myself, right?

**Reply by airtower-luna** [reply]
Yep, that's correct.

**Reply by askibinski** [reply]
In Github actions, see the `path` property for this.

**Reply by edison12a** [reply]
@askibinski Thanks!

**Comment by rungrotArunon** [comment]
Thanks you bro's

**Comment by rungrotArunon** [comment]
I can not do you community

[[Rungrot_Saibutdee]]

ในวันที่ จ. 28 พ.ย. 2022 22:55 น. Albert Skibinski ***@***.***>
เขียนว่า:

**Comment by edison12a** [comment]
Final Solution: 
Use github actions to deploy your pages. It lets you customize everything. 

( Reported by @airtower-luna )

**Comment by SebastianTroy** [comment]
This would be super useful because I could place my files inside a directory with the same name as my repo, allowing me to migrate my website without changing every single in-site href to account for the new root directory name...

**Comment by rungrotArunon** [comment]
Thank you for your compliment and I'm glad it's useful.

**Comment by kkarhan** [comment]
Anyone else think that this entire issue shouldn't exist and instead, GitHub Pages should just allow one to select *any arbitrary folder* instead?

Cuz it's really annoying and I just don't want to deal with GitHub Actions
![github-pages-error](https://github.com/user-attachments/assets/5fdf624a-7357-4c80-ba9f-b16de72209ca)
 at all!

Why even is there a menu then if they decide to only support 3 options at all...

**Reply by kkarhan** [reply]
The [GitHub Pages Overwirter]( https://github.com/marketplace/actions/github-pages-overwriter ) does work, but I still am mad about this frustration!

- Despite it being [successfully working](http://docs.os1337.com/)!

**Reply by airtower-luna** [reply]
You can [us an actions workflow to publish your site](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#creating-a-custom-github-actions-workflow-to-publish-your-site), that way you can choose whatever path you want. A workflow that only publishes static files can be pretty simple: https://github.com/actions/starter-workflows/blob/main/pages/static.yml

**Comment by steveniles** [comment]
It kind of amazes me that this is still an issue...

What's being asked for here, is to be able to be able to make a simple, (for many people) one-time change to their repo settings, so the pages are hosted from another folder (that isn't precisely, exactly named "docs").

This is akin to asking for a hammer to deal with a nail that's sticking out.

And the recurring solution offered, to use Github Actions, is akin to responding "sorry, we don't have a hammer, just learn how to use a jackhammer".

As has been said, it's overkill for this use-case. Come on Github, you've already built quite a complex system here, right down to the drop-down box that already exists for selecting "root" or "docs". Just finish the 99% done job and let people pick another folder as their pages base.

"When all you have is a jackhammer..."

**Comment by shaunagm** [comment]
Adding myself to the list of people who would greatly prefer an option in the settings dropdown. I can and do use github actions regularly but I'm also a teacher and having to teach people github actions can double the length of a session in my 'building a static site with github' training

**Reply by rayluo** [reply]
> ... I can and do use github actions regularly but I'm also a teacher and having to teach people github actions can double the length of a session in my 'building a static site with github' training

Until github provides their built-in solution to customize source folder, you may take a look at [Github Page Overwriter (GPO)](https://github.com/marketplace/actions/github-pages-overwriter) which requires minimal configuration, so it hopefully won't double the length of your teaching session.

P.S. I am also in the list of people waiting here, and I am the author of that GPO.

**Comment by darkoverlordofdata** [comment]
years later, still the same dump problems this is why i've left github

**Reply by JackGraymer** [reply]
wait, is this possible in another platform?
Example, can you choose the folder to be hosted in gitlab?

**Comment by diskria** [comment]
Honestly, this limitation makes no sense.
I’m building Gradle plugins and using GitHub Pages as a static Maven repository — it works perfectly fine technically, except that I’m forced to put everything inside the docs folder.
GitHub already lets us pick any branch in the UI, so why is the folder name hard-coded?

There’s no real technical reason for this — it just adds friction for anyone who uses Pages for something slightly more advanced than a static HTML site.
At this point, it feels like a fixable design oversight that’s been ignored for years.

**Reply by rayluo** [reply]
I don't know Gradle/Maven but, out of curiosity, would [Github Page Overwriter (GPO)](https://github.com/marketplace/actions/github-pages-overwriter) help for your case? (Shameless plug: I am the author of that GPO tool.)

**Comment by darkoverlordofdata** [comment]
Another reason to switch to gitlab


Bruce Davidson

**Comment by JackGraymer** [comment]
Adding myself to the list of people that would like this feature implemented.
Have your repo as you normally would, plus making a folder specifically for the site, with whatever name.
